### PR TITLE
Complex type support

### PR DESCRIPTION
**Pull request by Randl**
Rebase of https://github.com/patrick-kidger/diffrax/pull/197

### CODE REVIEW

File: diffrax/integrate.py [hunk 1]
```diff
@@ -654,6 +667,7 @@ def diffeqsolve(
     t1 = jnp.asarray(t1, dtype=dtype)
     if dt0 is not None:
         dt0 = jnp.asarray(dt0, dtype=dtype)
+    timelikes.append(dtype)
```
Comment (patrick-kidger) [inline diffrax/integrate.py:670]: Nit: this line is unnecessary, as `timelikes` isn't used after this point.

File: diffrax/nonlinear_solver/base.py [hunk 2]
```diff
@@ -106,4 +106,6 @@ def jac(fn: Callable, x: PyTree, args: PyTree) -> LU_Jacobian:
         if not jnp.issubdtype(flat, jnp.inexact):
             # Handle integer arguments
             flat = flat.astype(jnp.float32)
-        return jsp.linalg.lu_factor(jax.jacfwd(curried)(flat))
+        return jsp.linalg.lu_factor(
+            jax.jacfwd(curried, holomorphic=jnp.iscomplexobj(flat))(flat)
+        )
```
Comment (patrick-kidger) [inline diffrax/nonlinear_solver/base.py:111]: Given the guard against implicit solvers earlier, is this change necessary?

(I think we'd want a lot of tests to be sure that the nonlinear solvers handle complex numbers correctly, what with all the linear solves and such going on.)

**Review by patrick-kidger — COMMENTED** [review]
I think there's a couple of unnecessary changes. Other than that, LGTM!

File: diffrax/nonlinear_solver/base.py (hunk 2 above)
Comment (Randl) [reply diffrax/nonlinear_solver/base.py:111]: With this change, implicit solvers are working for me locally. If anything, I'd remove the guard?

Comment (patrick-kidger) [reply diffrax/nonlinear_solver/base.py:111]: So as it happens, I'm currently in the process of upgrading Diffrax to use [Lineax](https://github.com/google/lineax) and [Optimistix](https://github.com/patrick-kidger/optimistix) to handle the linear and nonlinear solving. This is going to be great from a feature point of view, but it does also mean that any changes we make here will be short-lived.

Other than that, I think we'd need some fairly comprehensive tests to check that things are working. There's subtleties here that can easily go wrong, e.g. transpose vs conjugate-transpose. (JAX does the former, PyTorch does the latter.)

To justify the above, btw -- the position I've been taking with the JAX sciML ecosystem is that it's better not to do something than it is to do it unreliably. Hence why complex support has been such a slow-moving thing to have happen!

Comment (Randl) [reply diffrax/nonlinear_solver/base.py:111]: I agree with you regarding taking slow part. But this code otherwise will just crash, and it passes basic tests. If you insist I'll remove it, or reinstate the guard

Comment (patrick-kidger) [reply diffrax/nonlinear_solver/base.py:111]: Yup, I'd reinstate the guard for now. (As I say, the current implementation will be deleted in ~2 weeks anyway.)
I think getting complex support into Lineax will be the main thing needed to get proper support here.

Comment (Randl) [reply diffrax/nonlinear_solver/base.py:111]: ok I've put it back

**Comment by patrick-kidger** [comment]
Closing in favour of #330, which fixes the pre-commit failure present here.

Thank you to all of @TimonHoess @Randl @andyElking for your efforts getting this in!
